src/lib/zkstring.ts

```diff
@@ -0,0 +1,148 @@
+import { Bool, Circuit, Field, Poseidon } from '../snarky';
+import { CircuitValue, prop } from './circuit_value';
+
+export class Character extends CircuitValue {
+  @prop value: Field;
+
+  constructor(value: Field) {
+    super();
+
+    if (Character.validateInput(value)) {
+      this.value = value;
+    } else {
+      throw Error('Character initialization did not pass validation');
+    }
+  }
+
+  equals(char: Character): Bool {
+    return this.value.equals(char.value);
+  }
+
+  toField(): Field {
+    return this.value;
+  }
+
+  toString(): string {
+    const charCode = Number(this.value.toString());
+    return String.fromCharCode(charCode);
+  }
+
+  static fromString(str: string) {
+    const char = Field(str.charCodeAt(0));
+    return new Character(char);
+  }
+
+  /* 
+  TODO: Add support for more character sets
+  ASCII reference: https://en.wikipedia.org/wiki/ASCII#Printable_characters
+  */
+  static validateInput(value: Field): Boolean {
+    if (Circuit.inProver() || Circuit.inCheckedComputation()) {
+      const ret = Bool.and(
+        value.gte(Field(32)),
+        value.lte(Field(126))
+      );
+
+      return ret.toBoolean();
+    }
+
+    return true;
+  }
+};
+
+export class ZKString extends CircuitValue {
+  @prop values: Character[];
+
+  constructor(values: Character[]) {
+    super();
+
+    this.values = values;
+  }
+
+  append(str: ZKString): ZKString {
+    return new ZKString([...this.values, ...str.values])
+  }
+
+  // returns TRUE if str is found in this zkstring
+  contains(str: ZKString): Bool {
+    if (Circuit.inProver() || Circuit.inCheckedComputation()) {
```

Comment:
Let's not discriminate between in-snark and out-of-snark like this. We want operations to work and be the same in both scenarios.

Reply:
@mitschabaude do you mean I shouldn't do this if statement?  I was getting errors trying to run `gt` outside of a checked computation.  I'm not sure about this specific circumstance, but I was getting errors in general so added this line basically everywhere.

Or do you just mean I should replicate the same behavior in or out?

Reply:
I think `.gt` is not needed here, because you don't have to constrain the lengths in the circuit if they are static. So you can remove the if statement, and the length check (or check the length in pure JS if you want to handle the error case). In any case, the behavior should be the same in & out